osu.Framework/Audio/Track/TrackLoudness.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using ManagedBass;
+using ManagedBass.Loud;
+using osu.Framework.Audio.Callbacks;
+using osu.Framework.Extensions;
+using osu.Framework.Logging;
+
+namespace osu.Framework.Audio.Track
+{
+    /// <summary>
+    /// Measures loudness of audio samples.
+    /// </summary>
+    public class TrackLoudness : IDisposable
+    {
+        private Stream? data;
+
+        private readonly Task readTask;
+
+        private readonly CancellationTokenSource cancelSource = new CancellationTokenSource();
+
+        private float? integratedLoudness;
+
+        /// <summary>
+        /// Measures loudness from provided audio data.
+        /// </summary>
+        /// <param name="data">
+        /// The sample data stream.
+        /// The <see cref="TrackLoudness"/> will take ownership of this stream and dispose it when done reading track data.
+        /// If null, loudness won't get calculated.
+        /// </param>
+        public TrackLoudness(Stream? data)
+        {
+            this.data = data;
+
+            readTask = Task.Run(() =>
+            {
+                if (data == null)
+                    return;
+
+                if (Bass.CurrentDevice < 0)
+                {
+                    Logger.Log("Failed to measure loudness as no bass device is available.");
+                    return;
+                }
+
+                FileCallbacks fileCallbacks = new FileCallbacks(new DataStreamFileProcedures(data));
+
+                int decodeStream = Bass.CreateStream(StreamSystem.NoBuffer, BassFlags.Decode | BassFlags.Float, fileCallbacks.Callbacks, fileCallbacks.Handle);
+
+                if (decodeStream == 0)
+                {
+                    Logger.Log($"Bass failed to create a stream while trying to measure loudness: {Bass.LastError}");
+                    fileCallbacks.Dispose();
+                    return;
+                }
+
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(20000);
+
+                try
+                {
+                    int loudHandle = BassLoud.Start(decodeStream, BassFlags.BassLoudnessIntegrated | BassFlags.BassLoudnessAutofree, 0);
+
+                    if (loudHandle == 0)
+                    {
+                        Logger.Log($"Failed to start BassLoud: {Bass.LastError}");
+                        return;
+                    }
+
+                    while (Bass.ChannelGetData(decodeStream, buffer, buffer.Length) >= 0)
+                    {
+                    }
+
+                    float integratedLoudness = 1;
```

Comment:
After double checking my docs. The returned integratedLoudness will be -infinity if the track is complete silence. Meaning that it won't ever really exceed -1 under normal circumstances. I believe that if we're going to be associating success with less than 0, we might as well revert my changes earlier making the default value of integratedLoudness = 1. Since the reason I did that was that the if statements were checking greater than 0 and associating it with failure. whereas these if statements are checking less than.
```suggestion
                    float integratedLoudness = 0;
```